hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/mongodb/MongoBatchSqlConnectorTest.java

```diff
@@ -0,0 +1,358 @@
+            instance.getLoggingService().addLogListener(Level.FINE, lookForProjectAndFilterStep);
+        }
+
+        MongoCollection<Document> collection = database.getCollection(collectionName);
+        collection.insertOne(new Document("firstName", "Luke").append("lastName", "Skywalker").append("jedi", true));
+        collection.insertOne(new Document("firstName", "Han").append("lastName", "Solo").append("jedi", false));
+        collection.insertOne(new Document("firstName", "Anakin").append("lastName", "Skywalker").append("jedi", true));
+
+        execute("CREATE MAPPING " + collectionName
+                + " ("
+                + (includeIdInMapping ? " id VARCHAR external name _id, " : "")
+                + " firstName VARCHAR, "
+                + " lastName VARCHAR, "
+                + " jedi BOOLEAN "
+                + ") "
+                + "TYPE MongoDB "
+                + "OPTIONS ("
+                + "    'connectionString' = '" + connectionString + "', "
+                + "    'database' = '" +  databaseName + "' "
+                + ")");
+
+        String force = forceTwoSteps ? " and cast(jedi as varchar) = 'true' " : "";
+        assertRowsAnyOrder("select firstName, lastName from " + collectionName
+                        + " where lastName = ? and jedi=true" + force,
+                singletonList("Skywalker"),
+                asList(
+                        new Row("Luke", "Skywalker"),
+                        new Row("Anakin", "Skywalker")
+                )
+        );
+        assertEquals(forceTwoSteps, projectAndFilterFound.get());
+        for (HazelcastInstance instance : instances()) {
+            instance.getLoggingService().removeLogListener(lookForProjectAndFilterStep);
+        }
+    }
+
+
+    protected void execute(String sql, Object... arguments) {
+        try (SqlResult ignored = sqlService.execute(sql, arguments)) {
+            EmptyStatement.ignore(null);
+        }
+    }
+
+    @Test
+    public void readWithTypeCoertion() {
+        final String collectionName = methodName();
+
+        MongoCollection<Document> collection = database.getCollection(collectionName);
+        collection.insertOne(new Document("firstName", "Luke").append("lastName", "Skywalker").append("jedi", "true"));
+
+        createMapping(true);
+        collection.insertOne(new Document("firstName", "Han").append("lastName", "Solo").append("jedi", "false"));
+        collection.insertOne(new Document("firstName", "Anakin").append("lastName", "Skywalker").append("jedi", "true"));
+
+        assertRowsAnyOrder("select firstName, lastName, jedi from " + methodName()
+                        + " where lastName = ?",
+                singletonList("Skywalker"),
+                asList(
+                        new Row("Luke", "Skywalker", true),
+                        new Row("Anakin", "Skywalker", true)
+                )
+        );
+    }
+
+    @Test
+    public void readsWithJoins() {
```

Comment:
Maybe also a join with IMap?

Reply:
Good idea, d9b04b0c42b49f45f67614f6cb82e2984b0ff081

Reply:
Maybe as a separate test and only join 2 mappings at a time.

Reply:
95d588f0f37cf4ff88edff5f638e4f2bcc9fa7f6